### PR TITLE
feat: add explicit experiment phases with atomic counter and 80% sampling threshold

### LINKED ISSUE #132: Add explicit experiment phases with atomic counter and 80% sampling threshold
## What to build

Replace the implicit single-phase experiment execution with an explicit state machine and decouple sampling from evaluation.

**Experiment state machine:** `pending → sampling → evaluating → completed` (or `failed`).

**`pending_samples_count` counter:** initialised to `dataset.dataset_samples.count` when the experiment transitions to `sampling`. Each sampling job decrements it atomically under `with_lock` on success or when retries are exhausted (via `retry_on` exhausted block in Active Job).

**Phase transition logic (inside each sampling job, after decrementing):**
When `pending_samples_count` reaches zero, check: if `completed_samples >= total * 0.8`, transition to `evaluating` and enqueue one evaluation job per `Sample`; otherwise transition to `failed`.

**Job split:**
- `ExperimentJob` transitions experiment to `sampling`, sets `pending_samples_count`, enqueues one `SamplingJob` per `DatasetSample`
- `SamplingJob` calls `Evaluation::Sampler`, decrements counter, triggers phase transition when counter hits zero
- `EvaluationJob` calls `LLMJudgeEval` for one `Sample`, marks experiment `completed` when all evaluation jobs are done (same atomic counter pattern)

## Acceptance criteria

- [ ] Experiment progresses through `pending → sampling → evaluating → completed` for the happy path
- [ ] Experiment transitions to `failed` if fewer than 80% of samples complete successfully
- [ ] `pending_samples_count` is decremented atomically; no race conditions under concurrent jobs
- [ ] A failed sampling job that exhausts retries decrements the counter and triggers the threshold check
- [ ] Specs cover the 80% threshold boundary (exactly 80% passes, below 80% fails)

## Blocked by

- #130

### PR DESCRIPTION
**Pull request by alaexeyshustov**
## Summary
- Replace implicit single-phase run with AASM state machine: `pending → sampling → evaluating → completed / failed`
- Split `RunEvalJob` into `SamplingJob` + `EvaluationJob`, each decrementing an atomic counter (`pending_samples_count` / `pending_evaluations_count`) under `with_lock`
- Phase transition after sampling: if ≥ 80% of samples completed, advance to `evaluating`; otherwise transition to `failed`

Closes #132

## Test plan
- [x] AASM state machine transitions (9 model specs)
- [x] `ExperimentJob` transitions to sampling, sets counter, enqueues `SamplingJob` per dataset sample
- [x] `SamplingJob` threshold boundary: exactly 80% passes, below 80% fails
- [x] `SamplingJob` retry exhaustion decrements counter and triggers threshold check
- [x] `EvaluationJob` decrements counter and marks experiment completed when all done
- [x] Retry exhaustion in `EvaluationJob` still completes the experiment
- [x] Full suite: 1903 examples, 0 failures, 96.64% coverage
- [x] Rubocop: 0 offenses

🤖 Generated with [Claude Code](https://claude.com/claude-code)
   Tokens used: input=75, output=64788, cache_read=6443025, cache_write=120945

### CODE REVIEW

**Comment by alaexeyshustov** [comment]
## Code Review

### Security
**No security vulnerabilities found.** The whitelist-before-`constantize` pattern in both jobs is correctly implemented, and `evaluator_classes` is hardcoded in all code paths (never user-input).

---

### Functionality & Performance

**CRITICAL — TOCTOU race in `ExperimentJob` between `start_sampling!` and `update!(pending_samples_count:)`** (`experiment_job.rb:10-12`)
Between transitioning to `sampling` and writing the counter, a `SamplingJob` could be enqueued and decrement before the counter is initialised. Fix: set `pending_samples_count` first, then call `start_sampling!`.

**CRITICAL — Empty dataset hangs in `sampling` state** (`experiment_job.rb:11-13`)
If `dataset.dataset_samples.count == 0`, `pending_samples_count` is set to 0 but no `SamplingJob` is ever enqueued and no threshold check fires — the experiment never exits `:sampling`. Fix: transition to `:failed` (or skip straight to `:evaluating` with 0 evaluation jobs) when `total == 0`.

**HIGH — N+1 query in `evaluate_threshold`** (`sampling_job.rb:41-45`)
`experiment.samples.count` and `experiment.samples.ids` are two separate queries inside the lock. Fetch IDs once and derive the count from the result: `ids = experiment.samples.ids; return ids.size >= total * 0.8 ? (set counter; transition; ids) : (fail; nil)`.

**MEDIUM — Negative counter possible on retry after a restart** (`sampling_job.rb:33`)
If `ExperimentJob` is re-run while retries are still in-flight, the retry's exhaustion block decrements a freshly-reset counter. Guard with: `break if experiment.pending_samples_count <= 0` before `decrement!`.

---

### Code Quality & Readability

**HIGH — `retry_on` block calls private method via `send`** (`sampling_job.rb:9`, `evaluation_job.rb:9`)
`job.send(:decrement_and_transition, experiment)` bypasses encapsulation and makes the retry path invisible in stack traces. Extract the shared decrement logic to a service object or make the method public with a clear domain name.

**MEDIUM — `evaluate_threshold` returns `nil` on the failure branch** (`sampling_job.rb:50`)
Returning `nil` forces the `sample_ids&.each` guard at the call site. Return an empty array (`[]`) for a uniform, idiomatic contract.

**MEDIUM — Duplicated `evaluators_for` method** (`evaluation_job.rb:34-40`)
Identical whitelist-and-constantize logic existed in the now-deleted `RunEvalJob`; it is now duplicated in `EvaluationJob`. Extract to a shared module or plain service to keep it DRY.

**LOW — Missing spec for empty-dataset edge case**
No test for `dataset.dataset_samples.count == 0`. Add a spec that asserts the experiment does not hang in `:sampling`.

---

### Summary — Top 3 actionable items

1. **Fix TOCTOU in `ExperimentJob`**: write `pending_samples_count` _before_ calling `start_sampling!`, and add a guard for `total == 0`.
2. **Replace `send(:private_method)` in retry blocks**: extract the decrement-and-transition logic into a public or module-level method so the retry path is transparent.
3. **Collapse double query to single `ids` fetch** inside `evaluate_threshold` to eliminate the N+1 and make the 80%-check a single DB round-trip.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by alaexeyshustov** [comment]
Addressed all review comments in 8b3b50c:

**CRITICAL — TOCTOU in ExperimentJob**: `pending_samples_count` is now written _before_ `start_sampling!` so no `SamplingJob` can decrement an uninitialised counter.

**CRITICAL — Empty dataset hangs**: When `total == 0`, `ExperimentJob` now immediately calls `experiment.fail!` after transitioning to `sampling`. New spec added.

**HIGH — N+1 in `evaluate_threshold`**: Replaced the `samples.count` + `samples.ids` double-query with a single `experiment.samples.ids` call; count derived from `.size`.

**MEDIUM — Negative counter on retry**: Added `break if experiment.pending_samples_count <= 0` guard inside the lock before decrementing.

**HIGH — `send(:private_method)` in retry blocks**: `decrement_and_transition` (`SamplingJob`) and `decrement_and_maybe_complete` (`EvaluationJob`) are now public — retry blocks call them directly via `job.method_name`.

**MEDIUM — `evaluate_threshold` returning `nil`**: Now returns `[]` on the failure branch; the `&.each` guard at the call site is replaced with plain `.each`.

**MEDIUM — Duplicated `evaluators_for`**: `RunEvalJob` is already deleted so there is no live duplication; no change needed.

**LOW — Missing empty-dataset spec**: Added `"transitions the experiment to failed when the dataset has no samples"` to `experiment_job_spec.rb`.

File: app/jobs/evaluation/experiment_job.rb [hunk 1]
```diff
@@ -5,11 +5,13 @@ class ExperimentJob < ApplicationJob
...
+      return unless experiment.may_start_sampling?
 
-      experiment.update!(status: :running)
-      experiment.dataset.dataset_samples.each_with_index do |dataset_sample, index|
-        RunEvalJob.set(wait: 3.seconds * index).perform_later(experiment.id, dataset_sample.id)
+      experiment.start_sampling!
+      total = experiment.dataset.dataset_samples.count
+      experiment.update!(pending_samples_count: total)
+      experiment.dataset.dataset_samples.each do |dataset_sample|
+        SamplingJob.perform_later(experiment.id, dataset_sample.id)
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply app/jobs/evaluation/experiment_job.rb:30]: Fixed in a316963. The transition is now wrapped in `with_lock`: counter is set and `start_sampling!` called atomically. Added crash-recovery re-entry: if the job sees `sampling?` with `pending_samples_count == dataset_samples.count` (meaning transition succeeded but enqueuing never ran), it re-enqueues all sampling jobs. New spec covering this path added.

File: db/migrate/20260522000001_add_experiment_phases_to_evaluation_experiments.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    add_column :evaluation_experiments, :status, :string, default: "pending", null: false
+    add_column :evaluation_experiments, :pending_samples_count, :integer, default: 0, null: false
+    add_column :evaluation_experiments, :pending_evaluations_count, :integer, default: 0, null: false
+  end
+
+  def down
+    remove_column :evaluation_experiments, :pending_evaluations_count
+    remove_column :evaluation_experiments, :pending_samples_count
+    remove_column :evaluation_experiments, :status
+    add_column :evaluation_experiments, :status, :integer, default: 0
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply db/migrate/20260522000001_add_experiment_phases_to_evaluation_experiments.rb:30]: Fixed in a316963. The migration now adds a temporary `status_str` column, backfills it with a CASE statement mapping the old integer enum values (0→pending, 1→failed, 2→completed, 3→failed), then removes the original column and renames the new one. Existing rows are preserved rather than reset.

File: app/jobs/evaluation/sampling_job.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+module Evaluation
+  class SamplingJob < ApplicationJob
+    queue_as :default
+
+    retry_on StandardError, attempts: 3 do |job, _error|
+      experiment = Evaluation::Experiment.find_by(id: job.arguments[0])
+      next unless experiment
+
+      job.send(:decrement_and_transition, experiment)
+    end
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply app/jobs/evaluation/sampling_job.rb]: Fixed in a316963. Changed to `retry_on RubyLLM::Error` so only transient LLM/network errors (rate limits, server errors, overload) are retried. ArgumentError from misconfiguration, ActiveRecord::RecordNotFound, and other programming errors will now surface immediately.

File: app/jobs/evaluation/sampling_job.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    def evaluate_threshold(experiment)
+      total = experiment.dataset.dataset_samples.count
+      completed_count = experiment.samples.count
+
+      if completed_count >= total * 0.8
+        sample_ids = experiment.samples.ids
+        experiment.update!(pending_evaluations_count: sample_ids.size)
+        experiment.start_evaluating!
+        sample_ids
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply app/jobs/evaluation/sampling_job.rb:49]: Fixed in a316963. Changed to `sample_ids.size * 10 >= total * 8` (integer arithmetic) to eliminate any floating-point edge cases at exact 80% boundaries.

File: app/jobs/evaluation/evaluation_job.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    queue_as :default
+
+    ALLOWED_EVALUATORS = %w[Evaluation::Evaluators::LLMJudgeEval].freeze
+
+    retry_on StandardError, attempts: 3 do |job, _error|
+      experiment = Evaluation::Experiment.find_by(id: job.arguments[0])
+      next unless experiment
+
+      job.send(:decrement_and_maybe_complete, experiment)
+    end
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply app/jobs/evaluation/evaluation_job.rb]: Fixed in a316963. Changed to `retry_on RubyLLM::Error` — same rationale as SamplingJob: only retries transient LLM failures, lets unexpected exceptions surface.

File: app/models/evaluation/experiment.rb [hunk 6]
```diff
@@ -9,10 +11,33 @@ class Experiment < ApplicationRecord
...
     serialize :evaluator_classes, coder: JSON, type: Array
     serialize :metadata, coder: JSON
 
+    aasm column: :status do
+      state :pending, initial: true
+      state :sampling
+      state :evaluating
+      state :completed
+      state :failed
+
```
[the remark this answers is not in the document]
Comment (alaexeyshustov) [reply app/models/evaluation/experiment.rb:23]: Fixed in a316963. Added `in_progress?` predicate to `Evaluation::Experiment` (returns `sampling? || evaluating?`) and updated both occurrences in `experiment_detail_component.html.erb` from `running?` to `in_progress?`. RBS signature updated accordingly.
